core/src/main/java/com/datastax/oss/driver/internal/core/config/cloud/CloudConfigFactory.java

```diff
@@ -225,22 +228,40 @@ protected TrustManagerFactory createTrustManagerFactory(
   @NonNull
   protected BufferedReader fetchProxyMetadata(
       @NonNull URL metadataServiceUrl, @NonNull SSLContext sslContext) throws IOException {
-    try {
-      HttpsURLConnection connection = (HttpsURLConnection) metadataServiceUrl.openConnection();
-      connection.setSSLSocketFactory(sslContext.getSocketFactory());
-      connection.setRequestMethod("GET");
-      connection.setRequestProperty("host", "localhost");
-      return new BufferedReader(
-          new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8));
-    } catch (ConnectException e) {
-      throw new IllegalStateException(
-          "Unable to connect to cloud metadata service. Please make sure your cluster is not parked or terminated",
-          e);
-    } catch (UnknownHostException e) {
-      throw new IllegalStateException(
-          "Unable to resolve host for cloud metadata service. Please make sure your cluster is not terminated",
-          e);
+    HttpsURLConnection connection = null;
+    int attempt = 0;
+    while(attempt < METADATA_RETRY_MAX_ATTEMPTS){
+      try {
+        connection = (HttpsURLConnection) metadataServiceUrl.openConnection();
+        connection.setSSLSocketFactory(sslContext.getSocketFactory());
+        connection.setRequestMethod("GET");
+        connection.setRequestProperty("host", "localhost");
+        attempt++;
+        // if this is the last attempt, throw
+        // else if the response code is not 200, retry
+        // else, throw
+        if (attempt != METADATA_RETRY_MAX_ATTEMPTS && connection.getResponseCode() != 200) {
+          try {
+            Thread.sleep(METADATA_RETRY_INITIAL_DELAY_MS);
+            continue;
+          } catch (InterruptedException interruptedException) {
+            Thread.currentThread().interrupt();
+            throw new IOException("Interrupted while waiting to retry metadata fetch", interruptedException);
+          }
+        }
+        return new BufferedReader(
+                new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8));
```

Comment:
We know that connection.getInputStream() can throw an IOException under various cases (including some non-200 response codes).  Presumably if we add checks above to make sure we only hit this code if we have a 200 response code we can avoid that issue... but any such code is probably a bit more brittle than we'd like.  While we're here it may be best to just wrap this line of code in it's own try/catch and either (a) return an IllegalStateException if we get an IOException (like we do in the other catch blocks here) or (b) fold any such error into our retry logic.

Note that this issue is really what [JAVA-3033](https://datastax-oss.atlassian.net/browse/JAVA-3033) is about.